### PR TITLE
Fix non-ASCII characters across corefx

### PR DESCRIPTION
**Pull request by stephentoub**
A few issues:
- There were some non-visible characters causing problems for Visual Studio.
- There were a bunch of Unicode characters, which we try to avoid by policy.

The only cases I left were those that were parts of developers' names in license headers.

### CODE REVIEW

File: src/System.Runtime.WindowsRuntime/src/System/IO/IStorageFolderHandleAccess.cs [hunk 1]
```diff
@@ -7,8 +7,8 @@
 
 namespace System.IO
 {
-    // Available in 14393 (RS1) and later
-    [ComImport]
+    // Available in 14393 (RS1) and later
```
Comment (safern) [inline src/System.Runtime.WindowsRuntime/src/System/IO/IStorageFolderHandleAccess.cs]: Is this space vs tab?

**Review by safern — APPROVED** [review]
I'm curious how did you find all these?

File: src/System.Runtime.WindowsRuntime/src/System/IO/IStorageFolderHandleAccess.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Runtime.WindowsRuntime/src/System/IO/IStorageFolderHandleAccess.cs]: No, actually.  It's three non-breaking spaces ('\u00A0') followed by one normal space ('\u0020').  I changed it to four normal spaces.

**Comment by stephentoub** [comment]
> I'm curious how did you find all these?

I was hoping StyleCop would have an analyzer for it, but it didn't.  So I just wrote some throwaway code to loop over every file and every character in every file :)  We also don't currently run analyzers on tests, which is where the majority of these show up, so the analyzer approach wouldn't (currently) have helped us all that much, anyway.

**Comment by safern** [comment]
> We also don't currently run analyzers on tests, which is where the majority of these show up, so the analyzer approach wouldn't (currently) have helped us all that much, anyway.

We should be able to add StyleCop and other analyzers to tests. I think we should do it in order to maintain clean sources in between sources and tests. Thoughts?

**Comment by stephentoub** [comment]
> We should be able to add StyleCop and other analyzers to tests. I think we should do it in order to maintain clean sources in between sources and tests. Thoughts?

Fine thing to do, though likely a non-trivial amount of work.

File: src/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs [hunk 2]
```diff
@@ -24,9 +24,9 @@ public abstract class BindingRestrictions
         [SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
         public static readonly BindingRestrictions Empty = new CustomRestriction(AstUtils.Constant(true));
 
-        private const int TypeRestrictionHash = 1227133513;      // 00100 1001 0010 0100 1001 0010 0100 1001₂
-        private const int InstanceRestrictionHash = -1840700270; // 01001 0010 0100 1001 0010 0100 1001 0010₂
-        private const int CustomRestrictionHash = 613566756;     // 10010 0100 1001 0010 0100 1001 0010 0100₂
+        private const int TypeRestrictionHash =                    0b_0100_1001_0010_0100_1001_0010_0100_1001;
+        private const int InstanceRestrictionHash = unchecked((int)0b_1001_0010_0100_1001_0010_0100_1001_0010);
+        private const int CustomRestrictionHash =                  0b_0010_0100_1001_0010_0100_1001_0010_0100;
```
Comment (bartonjs) [inline src/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs]: While I don't mind changing these into their 0b values, it doesn't seem related to the change description.  Is this from another cleanup you're working on?

Comment (stephentoub) [reply src/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs]: > it doesn't seem related to the change description

It actually was.  See those little `2`s at the end of the comments?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs [hunk 3]
```diff
@@ -431,7 +431,7 @@ public void WriteSerializedString(string str)
         /// The string is UTF16 encoded and prefixed by the its size in bytes.
         ///
         /// This final byte holds the value 1 if and only if any UTF16 character within the string has any bit set in its top byte,
-        /// or its low byte is any of the following: 0x01–0x08, 0x0E–0x1F, 0x27, 0x2D, 0x7F. Otherwise, it holds 0.
+        /// or its low byte is any of the following: 0x01-x08, 0x0E-x1F, 0x27, 0x2D, 0x7F. Otherwise, it holds 0.
```
Comment (bartonjs) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs]: was the leading zero intentionally deleted?  (I'd expect it to say "0x01-0x08" still, not "0x01-x08"

Comment (stephentoub) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs]: Nope.  That was a typo.  Will fix.

File: src/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs (hunk 2 above)
Comment (bartonjs) [reply src/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs]: I do now.  ::sheepish::

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs [hunk 4]
```diff
@@ -1025,7 +1025,7 @@ public void WriteSerializedString(string value)
         /// The string is UTF16 encoded and prefixed by the its size in bytes.
         ///
         /// This final byte holds the value 1 if and only if any UTF16 character within the string has any bit set in its top byte,
-        /// or its low byte is any of the following: 0x01–0x08, 0x0E–0x1F, 0x27, 0x2D, 0x7F. Otherwise, it holds 0.
+        /// or its low byte is any of the following: 0x01-x08, 0x0E-0x1F, 0x27, 0x2D, 0x7F. Otherwise, it holds 0.
```
Comment (bartonjs) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: lost a zero up here, too

File: src/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs]: :smile:

File: src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs [hunk 5]
```diff
@@ -45,7 +45,7 @@ internal static IEnumerable<EccTestData> EnumerateApplicableTests()
             CurveOid = Secp256r1OidValue,
             CurveEncodedOidHex = Secp256r1OidHexValue,
 
-            // Suite B Implementers Guide to FIPS 186-3,
+            // Suite B Implementers Guide to FIPS 186-3,
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs]: I'm curious what letter here was weird.  Was this a homoglyph or something?

File: src/System.Text.Json/tests/Serialization/TestClasses.cs [hunk 6]
```diff
@@ -1784,10 +1784,10 @@ public void Verify()
 
     public class ClassWithUnicodeProperty
     {
-        public int Aѧ { get; set; }
+        public int A\u0467 { get; set; }
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/TestClasses.cs]: I had no idea that was legal C#.  Nifty.

**Review by bartonjs — APPROVED** [review]
LGTM once the two zeros are put back

File: src/System.Linq.Expressions/tests/ExceptionHandling/ExceptionHandlingExpressions.cs [hunk 7]
```diff
@@ -1073,10 +1073,8 @@ public void TryCatchThrowingWithinFilterCompiled()
         }
 
         private bool MethodWithManyArguments(
-            int x, int y, int z,
-            int α, int β, int γ, int δ,
-            int klaatu, int barada, int nikto,
-            int anáil, int nathrach, int ortha, int bháis, int @is, int beatha, int @do, int chéal, int déanaimh,
+            int a, int b, int c, int d, int e, int f, int g, int h, int i, int j,
```
Comment (krwq) [inline src/System.Linq.Expressions/tests/ExceptionHandling/ExceptionHandlingExpressions.cs]: I think this one might have been intentional

File: src/System.Private.Xml/tests/Xslt/XsltCompiler/CommonScenarios/XsltcTestFile.cs [hunk 8]
```diff
@@ -46,9 +46,9 @@ public XsltcTestFile(ITestOutputHelper output) : base(output)
         [InlineData("@infft14.txt", "fft14.dll", "no", "fft14", "fft14.pdb", "no", "fft14.txt")]
         //[Variation("15", Desc = "Test multiple config files with circular reference and case sensitive file names specified", Pri = 1, Params = new object[] { "@infft15.txt", "fft15.dll", "no", "fft15", "fft15.pdb", "no", "fft15.txt" })]
         //[InlineData("@infft15.txt", "fft15.dll", "no", "fft15", "fft15.pdb", "no", "fft15.txt")] //Skipping this, it tries to load System.dll
-        //[Variation("17", Desc = "Exercise comments(#) within file", Pri = 1, Params = new object[] { "@infft17.txt", "fft17.dll", "yes", "fft17", "fft17.pdb", "no", "fft17.txt" })]
+        //[Variation("17", Desc = "Exercise comments(\u0091#\u0092) within file", Pri = 1, Params = new object[] { "@infft17.txt", "fft17.dll", "yes", "fft17", "fft17.pdb", "no", "fft17.txt" })]
```
Comment (krwq) [inline src/System.Private.Xml/tests/Xslt/XsltCompiler/CommonScenarios/XsltcTestFile.cs]: interesting, those don't show up on the left hand side

File: src/System.Linq.Expressions/tests/ExceptionHandling/ExceptionHandlingExpressions.cs (hunk 7 above)
Comment (stephentoub) [reply src/System.Linq.Expressions/tests/ExceptionHandling/ExceptionHandlingExpressions.cs]: For what purpose? It's part of an exception handling test.

File: src/System.Text.Json/tests/Serialization/TestClasses.cs (hunk 6 above)
Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/TestClasses.cs]: It was intentionally added to exercise our JSON escaping behavior when serializing :)

We should add one where all the characters (in a long name) are non-ascii.

File: src/System.Text.Json/tests/JsonStringTests.cs [hunk 9]
```diff
@@ -22,7 +22,7 @@ public static void TestDefaultCtor()
         [InlineData("new lines\r\n")]
         [InlineData("tabs\ttabs\t")]
         [InlineData("\\u003e\\u003e\\u003e\\u003e\\u003e")]
-        [InlineData("zażółć gęślą jaźń")]
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/JsonStringTests.cs]: @kasiabulat - FYI, for future testing :)

File: src/System.Text.Json/tests/Serialization/OptionsTests.cs [hunk 10]
```diff
@@ -254,12 +254,12 @@ public static IEnumerable<object[]> JsonEncodedTextStringsCustom
...
-                    new object[] { "ééééé\"êêêêê", "ééééé\\u0022êêêêê" },
-                    new object[] { "ééééé\\u0022êêêêê", "ééééé\\\\\\u0075\\u0030\\u0030\\u0032\\u0032êêêêê" },
-                    new object[] { "ééééé>>>>>êêêêê", "ééééé\\u003E\\u003E\\u003E\\u003E\\u003Eêêêêê" },
-                    new object[] { "ééééé\\u003e\\u003eêêêêê", "ééééé\\\\\\u0075\\u0030\\u0030\\u0033\\u0065\\\\\\u0075\\u0030\\u0030\\u0033\\u0065êêêêê" },
-                    new object[] { "ééééé\\u003E\\u003Eêêêêê", "ééééé\\\\\\u0075\\u0030\\u0030\\u0033\\u0045\\\\\\u0075\\u0030\\u0030\\u0033\\u0045êêêêê" },
+                    new object[] { "\u00E9\u00E9\u00E9\u00E9\u00E9\u00EA\u00EA\u00EA\u00EA\u00EA", "\u00E9\u00E9\u00E9\u00E9\u00E9\u00EA\u00EA\u00EA\u00EA\u00EA" },
+                    new object[] { "\u00E9\u00E9\u00E9\u00E9\u00E9\"\u00EA\u00EA\u00EA\u00EA\u00EA", "\u00E9\u00E9\u00E9\u00E9\u00E9\\u0022\u00EA\u00EA\u00EA\u00EA\u00EA" },
+                    new object[] { "\u00E9\u00E9\u00E9\u00E9\u00E9\\u0022\u00EA\u00EA\u00EA\u00EA\u00EA", "\u00E9\u00E9\u00E9\u00E9\u00E9\\\\\\u0075\\u0030\\u0030\\u0032\\u0032\u00EA\u00EA\u00EA\u00EA\u00EA" },
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/OptionsTests.cs]: This type of change does end up making harder to parse/reason about what the expected value should be (`\u` vs `\\u`):
https://github.com/dotnet/corefx/pull/39524#discussion_r303964442

Is there a cleaner way to format the non-ascii text to avoid this drawback of the guideline?

**Review by ahsonkhan — APPROVED** [review]

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs [hunk 11]
```diff
@@ -180,7 +180,7 @@ internal static byte[] DecodeX509SubjectKeyIdentifierExtension(byte[] encoded)
         {
             AsnReader reader = new AsnReader(encoded, AsnEncodingRules.BER);
             ReadOnlyMemory<byte> contents;
-            if (!reader.TryReadPrimitiveOctetStringBytes(out contents))
+            if (!reader.TryReadPrimitiveOctetStringBytes(out contents))
```
Comment (ahsonkhan) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs]: Was this detected as non-ascii whitespace?

File: src/System.Text.Json/tests/Serialization/OptionsTests.cs (hunk 10 above)
Comment (stephentoub) [reply src/System.Text.Json/tests/Serialization/OptionsTests.cs]: > Is there a cleaner way to format the non-ascii text to avoid this drawback of the guideline?

I'm not aware of one. The guideline exists because it's very easy to accidentally lose / mistransform non-ASCII characters when files roundtrip through various editors and text tools.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs (hunk 11 above)
Comment (stephentoub) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs]: Yes, it's value 160, a non-breaking space.

File: src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs]: It's value 146.

Comment (ahsonkhan) [reply src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs]: Looks like a custom back-tick within the "private use" range. Outlook shows it fine :)

image

https://www.fileformat.info/info/unicode/char/0092/index.htm

Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs]: Well, that explains why it looks like no change on the web :smile:

Comment (stephentoub) [reply src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs]: And VS doesn't show it at all:
image

This all speaks to why we don't like such characters in the code. :)

Comment (ahsonkhan) [reply src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs]: > Well, that explains why it looks like no change on the web 😄

Which browser? Chrome/Edge show the placeholder/replacement character.

Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/EccTestData.cs]: I see it like Steve's image in Firefox. (no visible character between r and s)

File: src/System.Text.Json/tests/Serialization/OptionsTests.cs (hunk 10 above)
Comment (krwq) [reply src/System.Text.Json/tests/Serialization/OptionsTests.cs]: @ahsonkhan HtmlDecode perhaps will do better in this test
